save mps features into db

ryanlau opened the pull request:
title

ryanlau committed:
save mps features into db

ryanlau committed:
correct path

SreeDan commented on cloud_webserver_v2/internal/mps/mps.go:
```diff
@@ -2,19 +2,34 @@ package mps
+	"go.mongodb.org/mongo-driver/bson/primitive"
 )
 
+type processingJob struct {
```
Maybe rename this `mpsJob`?

SreeDan commented on cloud_webserver_v2/internal/mps/mps.go:
```diff
@@ -2,19 +2,34 @@ package mps
 	// Contains the ids of all the jobs submitted and are being processed/queued
-	jobsProcessing []string
+	jobsProcessing []processingJob
+
```
any reason for extra line?

ryanlau replied:
no, removed

SreeDan commented on cloud_webserver_v2/internal/mps/mps.go:
```diff
@@ -2,19 +2,34 @@ package mps
+	jobsProcessing []processingJob
+
+	dbClient *database.DatabaseClient
+	ctx      context.Context
```
It's not really good practice to have context be a field in a struct. It muddies the waters of defining a context's lifetime and can make lifetimes a little unpredictable.  In general, context should be something explicitly passed into a function.

SreeDan commented on cloud_webserver_v2/internal/mps/mps.go:
```diff
@@ -90,7 +105,7 @@ func newMatlabJobRequestPayload(rhs []string) *matlabJobRequestPayload {
 }
 
-func NewMatlabClient(mpsBaseUrl string) *MatlabClient {
+func NewMatlabClient(ctx context.Context, dbClient *database.DatabaseClient, mpsBaseUrl string) *MatlabClient {
```
Instead of passing in a `ctx` here, it would make more sense to put the context's in each function,

SreeDan commented on cloud_webserver_v2/internal/mps/mps.go:
```diff
@@ -201,9 +242,22 @@ func (m *MatlabClient) deleteMatlabJobResult(jobId string) {
+
+	h5FilePath := model.MatFiles[0].FilePath
+
+	payload := newMatlabJobRequestPayload([]string{"/data/run_metadata/" + h5FilePath})
```
Let's make "/data/run_metadata" a const

ryanlau committed:
fix how context is passed

SreeDan commented on cloud_webserver_v2/internal/mps/mps.go:
```diff
@@ -135,8 +151,8 @@ func (m *MatlabClient) pollForResults() {
+				m.deleteMatlabJobResult(job.jobId)
 			} else {
 				newJobsProcessing = append(newJobsProcessing, job)
 			}
```
Lets put a delay on this, maybe 1 second or something

aesteri committed:
HT08->HT09

Uploaded files automatically get assigned HT08

considering HT08 does not exist anymore, it will be time efficient to auto-assign to HT09 since testing should ramp up

ryanlau committed:
Merge branch 'main' into mps-fix

ryanlau committed:
update mps job processing logic

ryanlau committed:
fix bugs and add logic for storing file results into s3

ryanlau committed:
.

ryanlau committed:
+gh action to build docker img

ryanlau committed:
make changes for gh action to work

ryanlau committed:
make changes for gh action to work

ryanlau committed:
don't crash if .env is missing

ryanlau committed:
don't crash if mps is offline

ryanlau committed:
fix

SreeDan commented on .github/workflows/create-image.yml:
```diff
@@ -0,0 +1,43 @@
+  push:
+    branches:
+      - main
+      - mps-fix
```
do we need `mps-fix` post merge?

ryanlau replied:
no, i will remove later

SreeDan commented on cloud_webserver_v2/internal/delivery/http/mcap_handler.go:
```diff
@@ -349,16 +350,18 @@ func (h *mcapHandler) UpdateMetadataRecordFromID(w http.ResponseWriter, r *http.
-			if runModel.MpsRecord == nil {
-				runModel.MpsRecord = make(map[string]interface{})
-			}
+			// mpsMetadata := make(map[string]interface{})
```
Pls remove unused comments

ryanlau replied:
oh, was gonna ask you abt this actually. 

we need to leave the commented code in if we wanted to be able to update MPS records with an HTTP request to the server. i personally don't want to have that functionality but what do you think

SreeDan commented on cloud_webserver_v2/internal/delivery/http/mcap_handler.go:
```diff
@@ -285,6 +285,11 @@ func (h *mcapHandler) ProcessMatlabJob(w http.ResponseWriter, r *http.Request) *
 	}
 	scripts := strings.Split(scriptsParam, ",")
 
+	versionParam := r.URL.Query().Get("version")
```
What is `version` in this context?

ryanlau replied:
the matlab scripts will all be packaged together into 1 package to be deployed onto MPS. version refers to the package version

SreeDan commented on cloud_webserver_v2/internal/delivery/http/mcap_handler.go:
```diff
@@ -285,6 +285,11 @@ func (h *mcapHandler) ProcessMatlabJob(w http.ResponseWriter, r *http.Request) *
 	}
 	scripts := strings.Split(scriptsParam, ",")
```
if `scripts` is referring to MATLAB scripts, we should rename it to be more descriptive.

ryanlau replied:
dy want me to rename the query param too or just the variable?

SreeDan commented on cloud_webserver_v2/internal/models/vehicle_run_model.go:
```diff
@@ -21,6 +21,22 @@ type FileModelResponse struct {
+// MpsScriptResultModel represents the schema of the returned result of a MATLAB script
+type MpsScriptResultModel struct {
+	// Type can be "mat", "image", or "text"
+	Type string `json:"type"`
```
It would make sense for `Type` and `Result` if had constants for values like `mat`, `image`, etc.

SreeDan commented on cloud_webserver_v2/internal/mps/mps.go:
```diff
@@ -171,11 +227,95 @@ func (m *MatlabClient) processResult(jobId string) {
+
+	switch scriptResult.Type {
+	case "mat", "image":
+		// scriptResult.Result = /data/mps_generated/file_name.mat
```
nit: unused comment

ryanlau replied:
oh i wanted to give an example of `scriptResult.Result` for future reference

SreeDan commented on cloud_webserver_v2/internal/mps/mps.go:
```diff
@@ -171,11 +227,95 @@ func (m *MatlabClient) processResult(jobId string) {
+			log.Fatalf("generated file does not exist: %s", mpsGeneratedFileLocation)
+		}
+
+		// copy the generated file to the local s3 cache directory
```
Can't we just simplify logic by just moving the file with`os.Rename`?

ryanlau replied:
bc we are moving files across docker volumes, `os.Rename` doesn't work so we have to unfortunately copy it over manually

SreeDan commented on cloud_webserver_v2/internal/mps/mps.go:
```diff
@@ -90,66 +142,70 @@ func newMatlabJobRequestPayload(rhs []string) *matlabJobRequestPayload {
-func (m *MatlabClient) pollForResults() {
+// Polls the MPS for the result of a job until it is ready
+// Once it's ready, it processes the job result and then deletes it off MPS
+func (m *MatlabClient) pollForJobResult(mpsJob mpsJob, s3Repo *s3.S3Repository) {
```
We may want to include a timeout here just in case.

ryanlau replied:
we can set a timeout on the MPS side.

so if a script doesn't finish by X time, it will cancel the execution and say that it timed out which can be handled here.

SreeDan commented on cloud_webserver_v2/internal/s3/s3_repository.go:
```diff
@@ -96,6 +98,38 @@ func (s *S3Repository) DeleteObject(ctx context.Context, bucket string, objectPa
+	}
+	defer outFile.Close()
+
+	_, err = io.Copy(outFile, resp.Body)
```
future improvement: if the object we want to download is very large we can have a for loop which copies `N` bytes at a time until we download the full file.

ryanlau replied:
sounds good

ryanlau committed:
use enum for MPSScriptResultType

ryanlau committed:
Merge branch 'main' into mps-fix

ryanlau committed:
update docker compose for new dockerfile
